src/credentials/DeviceAttestationConstructor.cpp

```diff
@@ -0,0 +1,173 @@
+    bool firmwareInfoExists             = false;
+    size_t vendorReservedIdx            = 0;
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    while ((error = tlvReader.Next()) == CHIP_NO_ERROR)
+    {
+        uint64_t tag = tlvReader.GetTag();
+
+        if (TLV::IsContextTag(tag))
+        {
+            switch (TLV::TagNumFromTag(tag))
+            {
+            case kCertificationDeclarationTagId:
+                VerifyOrReturnError(certificationDeclarationExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(certificationDeclaration));
+                certificationDeclarationExists = true;
+                break;
+            case kAttestationNonceTagId:
+                VerifyOrReturnError(attestationNonceExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(attestationNonce));
+                attestationNonceExists = true;
+                break;
+            case kTimestampTagId:
+                VerifyOrReturnError(timestampExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(timestamp));
+                timestampExists = true;
+                break;
+            case kFirmwareInfoTagId:
+                VerifyOrReturnError(firmwareInfoExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(firmwareInfo));
+                firmwareInfoExists = true;
+                break;
+            default:
+                return CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT;
+            }
+        }
+        else if (TLV::IsProfileTag(tag))
+        {
+            // vendor fields
+            bool seenProfile = false;
+            uint16_t currentVendorId;
+            uint16_t currentProfileNum;
+
+            currentVendorId   = TLV::VendorIdFromTag(tag);
+            currentProfileNum = TLV::ProfileNumFromTag(tag);
+            if (!seenProfile)
+            {
+                seenProfile = true;
+                vendorId    = currentVendorId;
+                profileNum  = currentProfileNum;
+            }
+            else
+            {
+                // check that vendorId and profileNum match in every Vendor Reserved entry
+                VerifyOrReturnError(currentVendorId == vendorId && currentProfileNum == profileNum,
+                                    CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+            }
+
+            ByteSpan vendorReservedEntry;
+            ReturnErrorOnFailure(tlvReader.Get(vendorReservedEntry));
```

Comment:
This requires the caller to have allocated large buffers with unknown possible size. This is why I recommend using `ContiguousBufferTLVReader::GetByteView`, which got added later than this PR got started.

Reply:
I agree, this was discussed.   Will investigate this.

Reply:
done